src/services/impl/ThreadPool.hh

```diff
@@ -0,0 +1,62 @@
+#ifndef RA_SERVICES_THREADPOOL_HH
+#define RA_SERVICES_THREADPOOL_HH
+#pragma once
+
+#include "services\IThreadPool.hh"
+
+#include <assert.h>
+#include <functional>
+#include <mutex>
+#include <queue>
+#include <thread>
+#include <vector>
+
+namespace ra {
+namespace services {
+namespace impl {
+
+class ThreadPool : public IThreadPool
+{
+public:
+    ~ThreadPool();
+
+    void Initialize(size_t nThreads) noexcept;
+
+    void RunAsync(std::function<void()>&& f) noexcept override
+    {
+        if (m_bShutdownInitiated)
+            return;
+
+        assert(!m_vThreads.empty());
+
+        {
+            std::unique_lock<std::mutex> lock(m_oMutex);
+            m_vQueue.push(std::move(f));
+        }
+
+        m_cvMutex.notify_one();
+    }
+
+    void Shutdown(bool bWait) noexcept override;
+
+    bool IsShutdownRequested() const noexcept override
+    {
+        return m_bShutdownInitiated;
+    }
+
+private:
+    void RunThread() noexcept;
+
+    std::vector<std::thread> m_vThreads;
+    std::queue<std::function<void()>> m_vQueue;
+    std::mutex m_oMutex;
+    std::condition_variable m_cvMutex;
```

Comment:
If you're not using a timer or moving the mutex around, locks should be std::scoped_lock.
`m_cvMutex `type should be `std::condition_variable_any`. 
If you still want it to be a `std::condition_variable `variant. You could just make it another mutex and put it in the scoped_lock. scoped_lock unlocks in reverse order as if `std::lock` was called (I don't think it's in VC++ yet though).

Reply:
this implementation seems to be the standard way to do notify/wait mutexes in c++. even the example in the `condition_variable_any` documentation page uses `unique_lock`: https://en.cppreference.com/w/cpp/thread/condition_variable_any/notify_one

Reply:
Seems good to me.